Vspd3

song50119 opened the pull request:
The PR process about VSPD
Related to [this PR](https://github.com/planetdecred/dcrlibwallet/pull/163)

raedah commented:
No need to have UTC on every line. Put it in the header  Date UTC

Trailing zeros can be truncated on the amounts. I believe wisdom already did this in the transaction details view.

raedah commented:
With detecting the terminal width, we should be able to center the QR code.

itswisdomagain commented:
> With detecting the terminal width, we should be able to center the QR code.

If we do this, we might get a weird UI on full screen mode because the account selection widget will be on the far left of the page while the rest of the page content sits in the horizontal middle.

Maybe we should just left align everything?

itswisdomagain commented:
I think this page needs a # column, would be especially helpful for wallets with high number of txs in history. Should confirm from @raedah though.
> ![image](https://user-images.githubusercontent.com/27733432/55873249-0428b500-5b87-11e9-91b0-ddb1711b6a5b.png)

Single account shouldn't have this tip. Should be something like `Hit ENTER to copy generated address, ESC to return to navigation menu`.
> ![image](https://user-images.githubusercontent.com/27733432/55842422-6e137100-5b2b-11e9-845d-7f923cb4af68.png)

Multiple account hint should be `Select account with ARROW DOWN and hit ENTER to generate address. Use ESC to return to navigation menu`.
> ![image](https://user-images.githubusercontent.com/27733432/55873180-dcd1e800-5b86-11e9-9131-e85eea58d3da.png)

The receive page should have a "generate new address" button beneath the address qr code and text to skip the previously generated unused address and generate a new one. Use `wallet.GenerateNewAddress()` instead of `wallet.ReceiveAddress()` for that purpose. Should discuss this with @raedah.

itswisdomagain commented on app/walletcore/helper.go:
```diff
@@ -32,7 +32,21 @@ func SimpleBalance(balance *Balance, detailed bool) string {
 // NormalizeBalance adds 0s the right of balance to make it x.xxxxxxxx DCR
 func NormalizeBalance(balance float64) string {
-	return fmt.Sprintf("%010.8f DCR", balance)
+	return fmt.Sprintf("%.3f DCR", balance)
```
Is this a necessary change for this PR?

itswisdomagain commented on app/walletcore/helper.go:
```diff
@@ -32,7 +32,21 @@ func SimpleBalance(balance *Balance, detailed bool) string {
+	if totalBalance != spendableBalance {
+	return fmt.Sprintf("Total %s (Spendable %s)", totalBalance.String(), spendableBalance.String())
+	} else {
+	return totalBalance.String()
```
go fmt please

itswisdomagain commented on terminal/pages/history.go:
```diff
@@ -109,22 +108,50 @@ func historyPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
 	return body
 }
 
+var displayedTxHashes []string
```
- Move this declaration to the top of the file, before the first function in this file
- In the `historyPage` function, clear this slice before transactions are loaded using `displayedTxHashes = []string{}`

itswisdomagain commented on terminal/pages/history.go:
```diff
@@ -109,22 +108,50 @@ func historyPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
 	}
 
+	loc, _ := time.LoadLocation("UTC")
+	currentDate := time.Now().In(loc).Add(1 * time.Hour)
```
Why are we adding 1 hour? A comment may be necessary here

itswisdomagain commented on terminal/pages/history.go:
```diff
@@ -109,22 +108,50 @@ func historyPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
+			return
+		}
+
+		transactionDate := time.Unix(tx.Timestamp, 0).In(loc).Add(1 * time.Hour)
```
Why are we adding 1 hour to current time and this time, only to subtract them later? it's like doing (a + 1) - (b + 1) which is essentially equivalent to doing a - b.

itswisdomagain commented on terminal/pages/history.go:
```diff
@@ -109,22 +108,50 @@ func historyPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
 
+	loc, _ := time.LoadLocation("UTC")
+	currentDate := time.Now().In(loc).Add(1 * time.Hour)
+	timeDifference, _ := time.ParseDuration("24h")
```
- `_24hours` instead of `timeDifference`, your code in subsequent lines will read better:
`if transactionDuration > _24hours` reads better than `if transactionDuration > timeDifference`

- `transactionDuration` is also a bad name, should be something like `transactionAge`, to show that the variable represents how old a transaction is, which is the time difference between now and when the transaction was created.

- can declare this variable at the top of this file as well: `var _24hours, _= time.ParseDuration("24h")`

itswisdomagain commented on terminal/pages/history.go:
```diff
@@ -109,22 +108,50 @@ func historyPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
+		}
+
+		transactionDate := time.Unix(tx.Timestamp, 0).In(loc).Add(1 * time.Hour)
+		transactionDuration := currentDate.Sub(transactionDate)
```
`transactionAge` instead of `transactionDuration`

itswisdomagain commented on terminal/pages/history.go:
```diff
@@ -109,22 +108,50 @@ func historyPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
+		transactionDate := time.Unix(tx.Timestamp, 0).In(loc).Add(1 * time.Hour)
+		transactionDuration := currentDate.Sub(transactionDate)
+	   
+	   	dateOutput := strings.Split(tx.FormattedTime, " ")
```
`txDateTime` instead of `dateOutput`

itswisdomagain commented on terminal/pages/history.go:
```diff
@@ -109,22 +108,50 @@ func historyPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
+	   	dateOutput := strings.Split(tx.FormattedTime, " ")
+
+	    if transactionDuration > timeDifference {
+	    	historyTable.SetCell(row, 0, tview.NewTableCell(fmt.Sprintln(dateOutput[0])).SetAlign(tview.AlignCenter))
```
`dateOutput[0]` can be used directly, no need to wrap in `fmt.Sprintln()`

itswisdomagain commented on terminal/pages/history.go:
```diff
@@ -109,22 +108,50 @@ func historyPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
+	currentDate := time.Now().In(loc).Add(1 * time.Hour)
+	timeDifference, _ := time.ParseDuration("24h")
+
+	var confirmations int32 
```
There really is no need to declare this variable, you should be able to use `walletcore.DefaultRequiredConfirmations` directly

itswisdomagain commented on terminal/pages/history.go:
```diff
@@ -109,22 +108,50 @@ func historyPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
+	    	historyTable.SetCell(row, 0, tview.NewTableCell(fmt.Sprintln(dateOutput[1])).SetAlign(tview.AlignCenter))
+		}		
+
+		if txns.Confirmations > confirmations{
```
- use `walletcore.DefaultRequiredConfirmations` instead of `confirmations`
- should be greater than or equal to

itswisdomagain commented on terminal/pages/overview.go:
```diff
@@ -16,18 +19,26 @@ func overviewPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, s
+	errorTextView := primitives.WordWrappedTextView("")
+	errorTextView.SetTextColor(helpers.DecredOrangeColor)
+
+	displayError := func(errorMessage string) {
```
If both balance and recent activity need to display an error, the error displayed first will be removed in order to add the second error. This is bad. What was wrong with the previous approach?

itswisdomagain commented on terminal/pages/overview.go:
```diff
@@ -16,18 +19,26 @@ func overviewPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, s
-		"Return to navigation menu with Esc")
+	hintTextView.SetText("TIP: Scroll recent activity table with ARROW KEYS. Return to navigation menu with ESC")
 
 	tabAndEscInputCapture := func(viewIndex int) func(event *tcell.EventKey) *tcell.EventKey {
```
This input capture function is longer necessary. Focus should be placed on the recent activity table, and that table alone should listen for esc key presses. Tabs should be ignored

itswisdomagain commented on terminal/pages/overview.go:
```diff
@@ -16,18 +19,26 @@ func overviewPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, s
+		overviewPage.AddItem(errorTextView, 2, 0, false)
+	}
+
+	balanceViews, balanceViewBoxes := renderBalanceSection(overviewPage, wallet, displayError)
```
These views and boxes were only needed to make moving between fields with tab possible. That's no longer necessary so `renderBalanceSection` and `renderRecentActivity` no longer need to return anything.

Since `renderBalanceSection` has been greatly simplified, you could merge both functions together and put the code here directly instead.

itswisdomagain commented on terminal/pages/overview.go:
```diff
@@ -67,143 +78,74 @@ func overviewPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, s
+	timeDifference, _ := time.ParseDuration("24h")
+
+	var confirmations int32 
+	confirmations = walletcore.DefaultRequiredConfirmations
```
A number of the review comments from the overview page applies here as well

itswisdomagain commented on terminal/pages/receive.go:
```diff
@@ -31,44 +31,69 @@ func receivePage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
-				}
-			}), 0, 1, true)
+		}), 2, 1, true)
+		qrCode := fmt.Sprintf(qr.ToSmallString(false))
```
`fmt.Sprintf` is not necessary

itswisdomagain commented on terminal/pages/receive.go:
```diff
@@ -31,44 +31,69 @@ func receivePage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
+		form := primitives.NewForm()
+		form.SetBorderPadding(0, 0, 0, 0)
+		form.SetHorizontal(true).
+		SetLabelColor(helpers.DecredLightBlueColor)
```
The screenshot doesn't show that this label color reflects. If it doesn't reflect, delete this line.

Sirmorrison replied:
it does form colors are default yellow

ademuanthony commented on terminal/pages/helpers.go:
```diff
@@ -0,0 +1,39 @@
+	return
+}
+
+func formatAmountDisplay(amount int64, maxDecimalPlaces int) string {
```
the display of amount has been advised to be in the format, x.xxxxxxxx DCR on other interfaces. `walletcore.NormalizeBalance` already returns this

itswisdomagain replied:
There is a different requirement from @raedah for terminal and cli.

ademuanthony commented on terminal/pages/history.go:
```diff
@@ -90,49 +97,72 @@ func historyPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
+	displayError func(errorMessage string)) {
+
+	// show a loading text at the bottom of the table so user knows an op is in progress
+	displayError("Fetching data...")
```
`Fetching data...` is not an error message

ademuanthony commented on terminal/pages/history.go:
```diff
@@ -90,49 +97,72 @@ func historyPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
+
+	// now format amount having determined the max number of decimal places
+	formatAmount := func(amount int64) string {
+		return formatAmountDisplay(amount, maxDecimalPlacesForTxAmounts)
```
we should use `walletcore.NormalizeBalance` to maintain consistency across all the interfaces

itswisdomagain replied:
There is a different requirement from @raedah for terminal and cli.

ademuanthony commented on terminal/pages/history.go:
```diff
@@ -90,49 +97,72 @@ func historyPage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
+			row := historyTable.GetRowCount()
+
+			historyTable.SetCell(row, 0, tview.NewTableCell(tx.FormattedTime).SetAlign(tview.AlignCenter))
+			historyTable.SetCell(row, 1, tview.NewTableCell(tx.Direction.String()).SetAlign(tview.AlignCenter))
```
I think will be better to have this left aligned

itswisdomagain replied:
The UI has already been approved by @raedah, would not want to make additional changes.

ademuanthony commented on terminal/pages/overview.go:
```diff
@@ -2,213 +2,115 @@ package pages
+
+	// now format amount having determined the max number of decimal places
+	formatAmount := func(amount int64) string {
+		return formatAmountDisplay(amount, maxDecimalPlacesForTxAmounts)
```
we can use `walletcore.NormalizeBalance` for consistency

itswisdomagain replied:
There is a different requirement from @raedah for terminal and cli.

ademuanthony commented on terminal/pages/overview.go:
```diff
@@ -2,213 +2,115 @@ package pages
+	}
 
 	for _, tx := range txns {
 		row := historyTable.GetRowCount()
```
`rowCount` i think

ademuanthony commented on terminal/pages/overview.go:
```diff
@@ -2,213 +2,115 @@ package pages
-		historyTable.SetCell(row, 5, tview.NewTableCell(tx.Type).SetAlign(tview.AlignCenter))
-		historyTable.SetCell(row, 6, tview.NewTableCell(tx.Hash).SetAlign(tview.AlignCenter))
+		historyTable.SetCell(row, 0, tview.NewTableCell(tx.FormattedTime).SetAlign(tview.AlignCenter))
+		historyTable.SetCell(row, 1, tview.NewTableCell(tx.Direction.String()).SetAlign(tview.AlignCenter))
```
I think this can be Left Aligned

ademuanthony commented on terminal/pages/receive.go:
```diff
@@ -13,71 +13,98 @@ import (
 
-func generateAddress(wallet walletcore.Wallet, accountNumber uint32) (string, *qrcode.QRCode, error) {
+func generateAddressAndQrCode(wallet walletcore.Wallet, accountNumber uint32) (string, *qrcode.QRCode, error) {
 	generatedAddress, err := wallet.ReceiveAddress(accountNumber)
```
You should call `wallet.GenerateNewAddress` instead

itswisdomagain replied:
`GenerateNewAddress` is only when user wants to ignore previously generated but unused addresses to get a new one instead, it's not the default behaviour. The default behaviour is to repeat previously generated address until it is used before a new one is generated. That's what `ReceiveAddress` does.

song50119 committed:
Add Ticket Page
* Add commands NewVSPD, TicketPrice and PurchaseTicket
* Add Ticket Page to navTab

song50119 committed:
Update ticket page
* Show tickets on page
* Adds window walletTickets field
* Adds command (*Wallet).GetAllTickets
* Adds reponse type Ticket, Tickets

song50119 committed:
Show error on the UI if CreateTicketFeeTx fail

song50119 committed:
Update dcrutil version to v3

song50119 committed:
Update to use dcrlibwallet at master
- go.mod use dcrlibwallet latest master
- implement listener OnCFiltersFetchProgress

song50119 committed:
Update Ticket Page
- Clear error messages when password modal closed
- Sync list ticket when block attached
- Use VSPD host remote instead local

oshorefueled committed:
fix lag on navigating to tickets page

oshorefueled committed:
automate ticket purchase and staking processes

automate ticket purchase and staking processes

* display ticket purchase widgets and ticket list on the same page
* display list of unconfirmed ticket purchases
* replace wallet and account tabs with dropdowns

initialize vspd before every ticket purchase
* add wallet balance widget works with page dropdowns

go fmt code

rebase on master

remove unnecessary else statement

oshorefueled committed:
move ticket filtering to commands

oshorefueled committed:
add error log to ticket purchase methods

oshorefueled committed:
update unconfirmed list on ticket purchase

update unconfirmed list on ticket purchase

fix loading page displaying after purchasing ticket

rebase on master

go import code

Sirmorrison commented on ui/tickets_page.go:
```diff
@@ -0,0 +1,477 @@
+						walletID := common.info.Wallets[pg.walletsDropdown.SelectedIndex()].ID
+						tickets := (*pg.tickets).Confirmed[walletID]
+						if len(tickets) == 0 {
+							return common.theme.Body2("No ticket").Layout(gtx)
```
`No tickets found` seems better.

oshorefueled replied:
More concrete UI definitions will be made when mockups are created. This PR is just to get the functionality working

Sirmorrison replied:
Took this from `decrediton`. I get that a more concrete UI will be made available but its nice we get some things right for a start.

Sirmorrison commented on ui/tickets_page.go:
```diff
@@ -0,0 +1,477 @@
+	)
+}
+
+// func (pg *ticketPage) ticketRowInfo(gtx layout.Context, c *pageCommon, ticket wallet.Ticket) layout.Dimensions {
```
Remove commented code.

Sirmorrison commented on ui/tickets_page.go:
```diff
@@ -0,0 +1,477 @@
+	c.closeModal()
+}
+
+func (pg *ticketPage) getTicketFee(ticketHash string, password []byte) (feeTransaction string, err error) {
```
This function is be getting and creating the ticket fee simultaneously. 
Also, the vspd methods used within it are not used anywhere else which makes the `getTicketFee()` not so useful as the methods can be called directly in `purchaseTicket()`.
Kindly remove the the `getTicketFee()` and call 
`resp, err := pg.vspd.GetVSPFeeAddress(ticketHash, password)` and `feeTransaction, err = pg.vspd.CreateTicketFeeTx(resp.FeeAmount, ticketHash, resp.FeeAddress, password)` directly in `purchaseTicket()`.

oshorefueled replied:
The method can be renamed to "createTicketFeeTransaction" since what it gets the fee, creates a transaction and returns it. However, methods or functions aren't just for reuse. They're also pretty useful for keeping code clean, easy to read and organized. As is evident in the layout code when we define sections as methods even though they might not be reused. That was the goal for putting ticket transaction calls in a separate method.

Sirmorrison replied:
Renaming should work.

Sirmorrison commented on wallet/commands.go:
```diff
@@ -801,3 +807,142 @@ func (wal *Wallet) SetupAccountMixer(walletID int, walletPassphrase string, errC
+		wal.Send <- resp
+	}()
+	return hash[0], nil
+	//return "", nil
```
Remove commented code.

Sirmorrison commented on ui/tickets_page.go:
```diff
@@ -0,0 +1,477 @@
+		unconfirmedList:      layout.List{Axis: layout.Vertical},
+		ticketPageContainer:  layout.List{Axis: layout.Vertical},
+		ticketPurchaseList:   layout.List{Axis: layout.Vertical},
+		inputNumberTickets:   common.theme.Editor(new(widget.Editor), "Enter tickets amount"),
```
`"Enter tickets amount"` should be `"Enter ticket(s) amount"`

Sirmorrison commented on ui/tickets_page.go:
```diff
@@ -0,0 +1,477 @@
+							gtx.Constraints.Min.X = gtx.Px(values.MarginPadding60)
+							return txt.Layout(gtx)
+						}),
+						layout.Rigid(func(gtx C) D {
+							gtx.Constraints.Min.X = gtx.Px(values.MarginPadding120)
+							txt.Text = "Date (UTC)"
+							return txt.Layout(gtx)
+						}),
+						layout.Rigid(func(gtx C) D {
+							gtx.Constraints.Min.X = gtx.Px(values.MarginPadding120)
+							txt.Text = "Status"
+							return txt.Layout(gtx)
+						}),
```
`values.MarginPadding120` repeated twice. Kindly declare this a variable and reuse. `m120 := values.MarginPadding120`
```suggestion
		layout.Rigid(func(gtx C) D {
			gtx.Constraints.Min.X = gtx.Px(m120)
			txt.Text = "Date (UTC)"
			return txt.Layout(gtx)
		}),
		layout.Rigid(func(gtx C) D {
			gtx.Constraints.Min.X = gtx.Px(m120)
			txt.Text = "Status"
			return txt.Layout(gtx)
		}),
```

Sirmorrison commented on ui/tickets_page.go:
```diff
@@ -0,0 +1,477 @@
+							return txt.Layout(gtx)
+						}),
+						layout.Rigid(func(gtx C) D {
+							gtx.Constraints.Min.X = gtx.Px(values.MarginPadding150)
+							txt.Text = "Hash"
+							return txt.Layout(gtx)
+						}),
+						layout.Rigid(func(gtx C) D {
+							gtx.Constraints.Min.X = gtx.Px(values.MarginPadding150)
+							txt.Text = "Amount"
+							txt.Alignment = text.End
+							return txt.Layout(gtx)
```
same goes for this one.

oshorefueled committed:
clean up code

clean up code

* resolve bug with only one ticket fee being paid for
*remove commented code

remove imported from list of accounts

Sirmorrison reviewed (approved):
PR tested for functionality and it looks good to go.
